xarray/core/dataset.py

```diff
@@ -419,6 +419,18 @@ def _replace_vars_and_dims(self, variables, coord_names=None,
             obj = self._construct_direct(variables, coord_names, dims, attrs)
         return obj
 
+    def _replace_indexes(self, indexes):
+        variables = OrderedDict()
+        for k, v in iteritems(self._variables):
+            if k in indexes.keys():
+                idx = indexes[k]
+                variables[k] = Coordinate(idx.name, idx)
```

Comment:
If `idx.name != k` above, then this could be constructing an invalid dataset.

I think we should create `Coordinate(k, idx)` and then remap back to the original names below, if necessary.